core/src/main/kotlin/com/strumenta/kolasu/ids/DeclarativeNodeIdProvider.kt

```diff
@@ -1,67 +1,145 @@
 import kotlin.reflect.full.isSuperclassOf
 
+// TODO merge into NodeIdProvider.kt file
+
 /**
- * Declarative node id provider instances allow to define
- * language-specific or custom identifier construction policies.
- *
- * ```kotlin
- * object MyDeclarativeNodeIdProvider: DeclarativeNodeIdProvider(
- *     idFor(PackageDeclaration::class) { it.node.name }
- *     idFor(ClassDeclaration::class) { (node: ClassDeclaration, identifiers: NodeIdProvider) ->
- *         val packageIdentifier = node.findAncestorOfType(PackageDeclaration::class)?.let { identifiers.idFor(it) } ?: "__none__"
- *         "${packageIdentifier}::${node.name}"
- *     }
- * )
- * ```
+ * Utility function to define a [NodeIdProvider] for a specific language.
+ * @param init the configuration rules for the [NodeIdProvider]
+ * @return a [NodeIdProvider] instance with the configured rules
  **/
-open class DeclarativeNodeIdProvider(
-    vararg rules: DeclarativeNodeIdProviderRule<out Node>
-) : NodeIdProvider {
-    private val rules: List<DeclarativeNodeIdProviderRule<out Node>> = rules.sorted()
-
-    override fun id(kNode: Node): String {
-        return this.rules.firstOrNull { it.canBeInvokedWith(kNode::class) }?.invoke(this, kNode)
-            ?: throw RuntimeException("Cannot find rule for node type: ${kNode::class.qualifiedName}")
-    }
+fun nodeIdProvider(
+    init: NodeIdProviderConfigurationApi.() -> Unit
+): NodeIdProvider {
+    return ConfigurableNodeIdProvider().apply(init)
 }
 
+@DslMarker
+annotation class NodeIdProviderDsl
+
 /**
- * Utility function to define declarative node id provider rules.
- * Can be used whenever listing the rules in the DeclarativeNodeIdProvider constructor.
+ * Defines the NodeId Provider Configuration Api
+ * to use while configuring a [NodeIdProvider] instance.
  **/
-inline fun <reified NodeTy : Node> idFor(
-    noinline specification: NodeIdProvider.(NodeTy) -> String
-): DeclarativeNodeIdProviderRule<NodeTy> = DeclarativeNodeIdProviderRule(NodeTy::class, specification)
+@NodeIdProviderDsl
+sealed interface NodeIdProviderConfigurationApi {
+    /**
+     * Define the [String] identifier for the given [nodeType]
+     * using the given [rule].
+     * @param nodeType the node type for which this rule applies
+     * @param rule the node id provider rule
+     **/
+    fun <NodeTy : Node> idFor(
+        nodeType: KClass<out NodeTy>,
+        rule: NodeIdProviderConfigurationRuleApi.(NodeIdProviderConfigurationRuleContext<out NodeTy>) -> String
+    )
+}
 
 /**
- * Class representing a single rule of a DeclarativeNodeIdProvider.
+ * Defines the API when defining a node id provider
+ * rule for a specific node type.
  **/
-class DeclarativeNodeIdProviderRule<NodeTy : Node>(
+@NodeIdProviderDsl
+sealed interface NodeIdProviderConfigurationRuleApi {
+    /**
+     * Add an information [message] during the rule execution.
+     * @param message the information message content
+     **/
+    fun info(message: String)
```

Comment:
Would it make sense to group these messages in an interface called Logging, or something similar, and then have this interface extend that one?

Reply:
I am fine with this, we could also just define a set of extension methods over `MutableList<Issue>` and use them with the rule argument `issues`.

Reply:
Ok, both works for me

Reply:
I will create a _IssueExtensions_ file in the _semantics_ module introducing such extension methods. As concerns those used in _DeclarativeNodeIdProvider_, if we keep it, the methods remain separated as those used in _SemanticProvider_ subclasses are creating _semantic_ issues by default. Does this make sense?

Reply:
> I will create a IssueExtensions file in the semantics module introducing such extension methods. 

Perfect

> As concerns those used in DeclarativeNodeIdProvider, if we keep it, the methods remain separated as those used in SemanticProvider subclasses are creating semantic issues by default. Does this make sense?

Makes sense